### PR TITLE
Trip Photos Feature

### PR DESCRIPTION
**Pull request by jmoran-cei**
Full implementation is functioning. Users can add photos and upload them. They can see their photos listed and select ones they want to remove; they can also view them full screen.

Finished:
- Trip Photo File Uploading
  - add multiple files at once
  - remove a pending file
  - clear all pending files
  - file size `pipe` for each file
- Trip Photos Display
  - photo list displays when photos exist
  - implemented ability to select photos and delete them
  - select all & clear selection button
  - UI adapts on hover and click of certain elements
  - added full view (photo slideshow like Facebook) for displaying entire photos
- General
  - appropriate error handling per pending file
  - progress messages changes based on user actions / status
  - UI adapts to keep user updated with what is happening or just happened
  - some styling that displays well for different screen sizes

Demo Video: **image to start**
[demothumbnail](https://user-images.githubusercontent.com/107073235/218479681-9ee524b4-8450-44fc-8fd7-52285e3ce9e2.mp4)

### CODE REVIEW

**Review by irwinhurstCEI — COMMENTED** [review]
Looks good.
@swhook52  Take another look at trip-photo.service.ts and the trip-photos folder to see if I missed anything.

File: src/app/trip-page/trip-photo-full-view/trip-photo-full-view.component.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  initTrip: TripPhoto = {
+    id: '',
+    tripId: 0,
+    fileName: '',
+    fileType: '',
+    path: '',
+    addedByUser: '',
+    alt: '',
+  }
+  @Input() displayFullView$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
```
Comment (swhook52) [inline src/app/trip-page/trip-photo-full-view/trip-photo-full-view.component.ts]: Passing a BehaviorSubject into a component is not a best practice. Depending on what you are doing with this, you should look to inject a service into the constructor that calls a method that would next onto your BehaviorSubject. You always want to protect your BehaviorSubjects. In almost all cases they should be private. It's ok to create small services that allow a few close components to interact with each other.

File: src/app/trip-page/trip-photos-list/trip-photos-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    fileName: '',
+    fileType: '',
+    path: '',
+    addedByUser: '',
+    alt: '',
+  }
+  @Input() tripPhotos$: Observable<TripPhoto[]> = of([]);
+  @ViewChildren('photoselect') private photos!: QueryList<ElementRef>;
+  progressMessage$: BehaviorSubject<string> = new BehaviorSubject('');
+  selectedCount$: BehaviorSubject<number> = new BehaviorSubject(0);
```
Comment (swhook52) [inline src/app/trip-page/trip-photos-list/trip-photos-list.component.ts:21]: BehaviorSubjects should be private. If you want to expose the observable part of that to the view you should create a public Observable from the BehaviorSubject with `asObservable`

**Review by swhook52 — CHANGES_REQUESTED** [review]

File: src/app/trip-page/trip-photos-list/trip-photos-list.component.ts (hunk 2 above)
Comment (jmoran-cei) [reply src/app/trip-page/trip-photos-list/trip-photos-list.component.ts:21]: I was able to get it to work with a private behavior subject in the trip photos service and using a method to update it and an observable to access the behavior subject value from components. However, I realized for my use case I was wrong in originally using behavior subjects for child to parent  & parent to child updates so I removed them and just used observables. 

sidenote: I was able to remove a bit of change detection, but not all: I tried different things to try to get it to update the `tripPhotos$` observable between components upon photo file upload/deletion but wasn't able to get it to update w/o calling `detectChanges()`.

**Comment by jmoran-cei** [comment]
I was able to successfully finish implementing my trip photos feature. For better visual I uploaded a demo vid.

File: src/app/trip-page/trip-photos-list/trip-photos-list.component.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+          photosToRemove.push(tripPhotoObj);
+        }
+      }
+
+      // delete photos via API
+      this.tripPhotoService
+        .deletePhotos(photosToRemove, photos[0].tripId)
+        .pipe(take(1))
+        .subscribe({
+          next: (updatedTrip) => {
```
Comment (swhook52) [inline src/app/trip-page/trip-photos-list/trip-photos-list.component.ts:118]: `next` and `error` are a style that we really shouldn't be using. use `catchError` for errors and just use the body of the subscribe for `next`

**Review by swhook52 — APPROVED** [review]

**Pull request merged**